[B+C] Add API for improved block targeting methods.  Adds BUKKIT-5581

desht opened the pull request:
#### [B+C] Improved LivingEntity#getTargetBlock() and related methods. Adds BUKKIT-5581
##### The Issue:

While the current system for getting the block targeted by an entity works, there are several significant improvements to be made:
- The current methods are deprecated due to the use of magic numbers; a Set<Integer> containing raw block ids to be considered transparent.  In addition, this system provides poor flexibility; it should be possible to use criteria other than the block ID (or material) to determine transparency.
- The current methods do not work properly when the player is targeting a block behind another block with a smaller hitbox.  E.g. where a sign is attached to wall, player.getTargetBlock() will always return the sign, even if the player is targeting the wall behind the sign.
- The current methods return the targeted block (mostly; see above), but don't say which block face was targeted, or the precise location on the block where the crosshairs are.
##### Justification for this PR:

This PR adds new methods to resolve all of the above issues.
##### PR Breakdown:

The following changes are made:
- ~~Add new BlockTargetResult class.  This simple class aggregates the information returned by the new targeting methods: the targeted block, the targeted block face, and the precise targeted location.~~ This is no longer in the interface.
- ~~Add new BlockTransparencyFilter interface.  This functional interface has one method: `boolean isTransparent(Block)`. The isTransparent() method implementation will be called by the new targeting methods and should return true if the given block is to be considered transparent, false otherwise.~~ Dropped in favour of using `Predicate<Block>` instead.
- ~~Add `Vector[] getHitbox()` method to the Block interface.  This returns a pair of Vector objects representing opposite corners of the block's hitbox (which may be significantly smaller than the block's full 1x1x1 volume).~~ Added several hitbox-related methods to the Block interface to retrieve the block's hitbox details.  These new methods are used by the targeting methods to fix the hitbox problem mentioned above, but it's possible that plugins may also find this information useful.
- Add four new methods to LivingEntity (the first two are direct replacements of their old counterparts which take a `Set<Integer>`): 
  - `public List<Block> getLineOfSight(Predicate<Block> filter, int maxDistance, boolean useHitBoxes)`
  - `public Block getTargetBlock(Predicate<Block> filter, int maxDistance, boolean useHitBoxes)`
  - `public BlockFace getTargetBlockFace(Block block, boolean useHitbox)`
  - `public Location getTargetLocation(Block block, boolean useHitbox)`
##### Testing Results and Materials:

A test plugin has been created.

Source: https://github.com/desht/testplugin
Binary: https://dl.dropboxusercontent.com/u/12467600/testplugin.jar

Instructions for use:
- For all tests, air, glass and white stained glass will be considered transparent (note: only white stained glass, not other colours - the old API couldn't do that!)
- Look at any block and type /target.  Messages will be printed to the console display detail about the targeted block as returned by the new `public Block getTargetBlock(Predicate<Block> filter, int maxDistance, boolean useHitBoxes)` method, and the associated `getTargetBlockFace()` and `getTargetLocation()` methods.
- With a bone in hand, right clicking while looking at any block (from any distance) will do the same as above.
- If the player is sneaking, hit boxes will be ignored and the entire 1.0x1.0x1.0 volume of blocks will be used (thus behaving more like the old methods)
- With a blaze rod in hand, left clicking any block will display the block's hitbox information to the console.
##### Relevant PR(s):

CB-1371 - https://github.com/Bukkit/CraftBukkit/pull/1371 - Implementation of new API
##### JIRA Ticket:

BUKKIT-5581 - https://bukkit.atlassian.net/browse/BUKKIT-5581

desht committed:
Add API for improved block targeting methods.  Adds BUKKIT-5581

Wolvereness commented on src/main/java/org/bukkit/block/Block.java:
```diff
@@ -381,4 +382,18 @@
+	 *
+	 * @return a pair of vectors
+	 */
+	Vector[] getHitBox();
```
Nope. There's gotta be a better way to represent this. I'd much prefer a `Tuple<Location, Vector>`.

desht replied:
Ideally this would be represented by a cuboid/AABB type of class, but Bukkit doesn't have one of those.

desht replied:
Where am I getting the Tuple class from?  I don't see any suitable implementation in Bukkit or included libraries.  As for the members, are you proposing that the Location represents the lower (X,Y,Z) corner of the hitbox, and the Vector the upper (X,Y,Z)?  Or would the Vector be the size of the hitbox?

ryantheleach replied:
What about stairs and half slabs?
I could see this being a lot more useful then just for your calculations if it were done properly.

desht replied:
Half slabs work exactly as expected:

```
[08:10:33 INFO]: *** Hitbox for block CraftBlock{chunk=CraftChunk{x=6z=21},x=109,y=72,z=339,type=STEP,data=0}
[08:10:33 INFO]:   109.0,72.0,339.0 => 110.0,72.5,340.0
```

Stairs are a little strange:

```
[08:10:50 INFO]: *** Hitbox for block CraftBlock{chunk=CraftChunk{x=6z=21},x=103,y=74,z=345,type=SMOOTH_STAIRS,data=5}
[08:10:50 INFO]:   103.5,74.5,345.5 => 104.0,75.0,346.0
```

Note that block.getHitbox() is just returning what the NMS Block class computes, given the block's type and data.  The actual hitbox computation for stairs is done in the NMS BlockStairs.class:

```
public void updateShape(IBlockAccess paramIBlockAccess, int paramInt1, int paramInt2, int paramInt3)
  {
    if (this.N)
      a(0.5F * (this.O % 2), 0.5F * (this.O / 2 % 2), 0.5F * (this.O / 4 % 2), 0.5F + 0.5F * (this.O % 2), 0.5F + 0.5F * (this.O / 2 % 2), 0.5F + 0.5F * (this.O / 4 % 2));
    else
      a(0.0F, 0.0F, 0.0F, 1.0F, 1.0F, 1.0F);
  }
```

My testing shows that stairs always seem to have a hitbox of (X+0.5,Y+0.5,Z+0.5)->(X+1,Y+1,Z+1), regardless of stair orientation (horizontal or vertical).  This does look a bit odd, but as I said, it's what the server is calculating.

Given the shape of stairs, hitbox calculation for them is inevitably going to be a little difficult; maybe a special-case is needed here where stairs always have a full-block hitbox.  The current targeting behaviour through stairs does work strangely; I haven't noticed this behavior with any other block, but I guess I'll need to do an exhaustive test...

desht replied:
OK, tested with pretty much every block (and various rotations) - stairs are the only problematic case that I can find.  And I suspect this is related to a vanilla problem which has been reported more than once: https://bugs.mojang.com/browse/MC-1307

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,54 @@
+	 *
+	 * @return the intersected location
+	 */
+	public Location getTargetLocation() {
```
Overkill?

desht replied:
Not really.   The intersection point is calculated anyway by the implementing code (it has to, if it's going to take block hitboxes into account), and this information could be useful to plugins.  For example, knowing precisely where a sign (or framed map) is clicked could be used to implement a basic touchscreen system.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTransparencyFilter.java:
```diff
@@ -0,0 +1,16 @@
+	 * @param block the block to test
+	 * @return true if the block should be considered transparent; false otherwise
+	 */
+	public boolean isTransparent(Block block);
```
Look up `Predicate<T>`, basically renders this class pointless.

desht replied:
Alright.  But you're in agreement that a predicate system makes more sense than a Set of Material or whatever?

Wolvereness replied:
Well, I can't concede that. However, the existence of this class is predominated by `Predicate<T>`.

desht replied:
My argument for `Predicate<Block>` over `Set<Material>`: by passing a predicate, the plugin gains the ability to determine transparency based on any property of the block.  E.g. my test plugin makes white stained glass transparent, but not other colours (admittedly using deprecated getData(), but one hopes a StainedGlass class will be added to Bukkit eventually).  With a `Set<Material>`, you're limited to testing the block's material type, nothing else.

Black-Hole commented:
How about a default implementation of BlockTransparencyFilter using Set<Material> and overloaded methods in LivingEntity with a Set<Material> parameter?

desht committed:
API update: ditch BlockTransparencyFilter, use Predicate<Block>

desht committed:
And actually remove BlockTransparencyFilter (oops)

desht commented:
@Black-Hole the BlockTransparencyFilter class has been dropped (see line notes above).  I guess an overloaded implementation is possible, but not sure of the value; it's easy just to pass in a null parameter, which has the same meaning as in the old methods (treat only air as transparent).

Update: maybe I misunderstood.  Are you suggesting an overloaded method which takes a `Set<Material>` instead of a `Predicate<Block>` and basically turns it into a predicate based on the block type's membership of the set?  E.g.:

```
public BlockTargetResult getTargetBlock(Set<Material> transparent, int maxDistance, boolean useHitBoxes) {
  return getTargetBlock(new Predicate<Block>() {
    @Override
    public boolean apply(Block block) {
       return transparent.contains(block.getType());
    }
  }, maxDistance, useHitBoxes);
}
```

That's a possible addition, but I'll leave it to @Wolvereness to make a judgement on that.  I suppose it would present a more familiar interface to those used to the old methods (but it would have to be a Set of Material, not Integer, of course).

spaceemotion commented on src/main/java/org/bukkit/entity/LivingEntity.java:
```diff
@@ -56,6 +58,26 @@
+	 * @return a list of BlockTargetResult objects, each of which represents one
+	 *     block along the entity's line of sight
+	 */
+	public List<BlockTargetResult> getLineOfSight(Predicate<Block> transparent, int maxDistance, boolean useHitBoxes);
```
This whole code is indented by tabs instead of spaces

desht replied:
Oops!  All fixed now.

desht committed:
replace tabs with spaces

me4502 commented:
+1

This change is amazing, I was asking about this exact thing just a few days ago.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,55 @@
+ * or
+ * {@link org.bukkit.entity.LivingEntity#getTargetBlock(com.google.common.base.Predicate, int, boolean)}
+ * or
+ * {@link org.bukkit.entity.LivingEntity#getLastTwoTargetBlocks(com.google.common.base.Predicate, int, boolean)}
```
1. Why is this fully qualified?
2. Pretty sure this is longer than 78 characters.

desht replied:
1. Intellij insists on the classes within the @link tag being fully qualified (or importing org.bukkit.entity.LivingEntity, which I could alternatively do).  I'm pretty sure this is correct behaviour; any linked class should be resolvable by the IDE and the Javadoc compiler.
2. Right, but it's a single {@link .. } tag - should it really be split over multiple lines?  Update: if both org.bukkit.entity.LivingEntity and com.google.common.base.Predicate are imported, the line length will be under 78 characters.  Your call.

Wolvereness replied:
Imports are preferred, but it's not intellij insisting that, it's how Javadoc works.

desht replied:
Imports it is, then.  I'll be pushing an update in a minute with some Javadoc fixes (including this one; also reducing the line length of some other comments and fixing a couple of minor typos)

desht committed:
Javadocs updates

glen3b commented:
@desht I believe that is what @Black-Hole is suggesting, and I think it would be a good idea. Client code that only wanted to filter on material type would not have to write a `Predicate<Block>` (which would be very similiar for each client implementation) for matching based on material types.

desht commented:
@glen3b yes, I can see where you're coming from.  It might be considered convenient if you have a large `Set<Material>`.  But reasons against:
- It could be considered API bloat.  It's another three methods added to the class to do something which can be done with the `Predicate<Block>`
- I'd argue that it's no more convenient than just using a `Predicate<Block>` anyway.  You have to construct/initialise the Set versus passing a Predicate with a method containing a single if, or maybe switch, statement.  (And I'm pretty sure that a switch on a enum type will be no less performant than a contains() operation on a Set of that enum type).

glen3b commented:
Like you said, I think we should leave the judgement to @Wolvereness. I think there are valid reasons to go with either way.

me4502 commented on src/main/java/org/bukkit/entity/LivingEntity.java:
```diff
@@ -86,6 +113,36 @@
+     *     the block.  If false, the block's full volume is used.
+     * @return a list of two BlockTargetResult objects
+     */
+    public List<BlockTargetResult> getLastTwoTargetBlocks(Predicate<Block> transparent, int maxDistance, boolean useHitBoxes);
```
What is the point of this, exactly? Why not make it just 'get all target block', and then the plugin can use the last two?

desht replied:
I added it since the old API has the method.  However, we could probably do without it.  I suspect the main reason for having the old method is to make it easy to determine the block face that is targeted, with `result.get(1).getFace(result.get(0))`.  But the new API provides the `BlockTargetResult#getTargetFace()` method to do that.  In addition, the old API method of determining the face doesn't work well if the player is extremely close to the target block (the returned list may have only a single element).

So yeah, there probably isn't a need for `getLastTwoTargetBlocks()`.

me4502 replied:
Thinking about it, a 'get all' method could be incredibly useful. For example, a weapon that destroys all blocks in a line.

glen3b replied:
@me4502 Are you referring to `getLineOfSight()`, which returns all blocks in the line of sight?

desht replied:
Yeah, `getLineOfSight()` will get you all the blocks along the line.

desht commented:
Just to summarise the current state of discussion, I believe the following points need addressing:
- The return type for `Block#getHitBox()`.  @Wolvereness has indicated that a 2-element array isn't acceptable and suggested a Tuple, but we don't have a Tuple class available.  Options:
  - Add a custom `HitBox` class
  - Add a more generic Cuboid/AABB class to Bukkit (could have a lot of applications)
  - Add a `Tuple<T,U>` class to Bukkit
  - Stick with a 2-element List or array
- Is `LivingEntity#getLastTwoTargetBlocks()` really necessary?
- Is `Predicate<Block>` the best way to define transparency?  I personally believe so, but a method taking a `Set<Material>` could be provided in addition.
- How to deal with the hitbox of stairs, which appears to be the only block with problematic hitbox data from the server.  My feeling here is just to use a whole-block hitbox.

nightpool commented:
@desht is http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html not in the apache commons library bukkit uses?
Does Bukkit not have a Cuboid class still? We should probably get one, but that seems out of scope for this commit.

I would vote for a `Predicate<Block>` especially because we can use `Predicates.in(Set<Block>)` and `Predicates.in(Sets.newHashSet(Block... e))` but it still provides flexibility for more complex transparency definitions.

glen3b commented:
@nightpool I believe that would be a practical solution, except that Bukkit uses Commongs lang 2.3 and that class was introduced in Lang 3.0. Also, `Predicates.in(Sets.newHashSet(Block... e))` will not work because you are checking (in the predicate) if the set contains a _particular block instance_, not that it has a specific material.

@desht My 2 cents: I think a cuboid class would be useful for Bukkit, it would, as you mentioned, allow for many applications, including mass block editing, entities within region, etc. This value could be returned by `Block#getHitBox()`, which I think would be best. However, this seems out of the scope of this PR.

As far as `LivingEntity#getLastTwoTargetBlocks()` is concerned, I think that really depends on the original use case. If, as you suggested, the intended use was code such as `result.get(1).getFace(result.get(0))`, then it has been obsoleted by `BlockTargetResult#getTargetFace()` and is not needed. I am inclined to agree with you that a `Predicate<Block>` is the way to go, as it is extremely flexible. However, I would also not mind an overload accepting a `Set<Material>`, as it provides API access similar to the old API.

desht commented:
@nightpool as @glen3b said, the tuple-related classes are in Commons Lang 3.0, which is in the server under net.minecraft.util, but not in Bukkit, so of no use to plugin developers.

me4502 commented:
I'm pretty sure they are still usable by plugins.

glen3b commented:
@me4502 I've tested it, and as I've already said, _Bukkit_ doesn't use Commons Lang 3.0, the version containing `Tuple`, therefore my code doesn't compile if I attempt to add a reference to it.

@desht I've looked under both the CraftBukkit and mc-dev repositories and been unable to find the `Tuple` class under `net.minecraft.server.util` (I haven't been able to find the package). Could you point me in the right direction, or is it not on the repositories online (just in decompiled MC source)?

desht commented:
@glen3b look in net.minecraft.util.org.apache.commons.lang3.tuple - there are Tuple-related classes in there (Pair, Triple, etc.).

@me4502 plugins would have to build against CraftBukkit.jar to use those classes, so that's a non-starter.

me4502 commented:
Why not update bukkit to 3.0, especially if mc uses it.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Represents the result of calling a method to determine blocks intersected
+ * by an entity's line of sight:
+ *
```
These javadocs are formatted horridly - remember that its primary purpose is to render in an emitter.

desht replied:
Format them as a HTML list, then?

Wolvereness commented on src/main/java/org/bukkit/block/Block.java:
```diff
@@ -381,4 +382,18 @@
+     *
+     * @return a pair of vectors
+     */
+    Vector[] getHitBox();
```
An array of vectors still makes me sad - it's extremely unintuitive.

desht replied:
Fair enough, but what's your suggestion?  You mentioned using a tuple, but there isn't a usable Tuple (Pair) class in Bukkit, only in CraftBukkit.  See also my state-of-the-PR post from May 7 regarding this.

Wolvereness replied:
Two methods, one for the _floor location_ (coordinates closest to negative infinity) (doesn't this... kind of already exist?), and one for the _size_ (a vector of positive coordinates), such that, `x * y * z ==`_`volume of entity`_.

If we had the tuple, that would be the two parts anyway, right?

Additionally, it may be nice to have a 'cached object' version of the method, such that, you pass the `Vector`/`Location` as a parameter, and it just updates (and returns) the object. Just like the `Location getLocation(Location)` in `Entity`.

desht replied:
Yeah, that works quite nicely.

Regarding the floor location: no, it doesn't exist in Bukkit already - consider for example the hit box of a flower.  Its floor location's Y co-ordinate is the block's base Y co-ordinate, but the X and Z co-ordinates are somewhere in the middle of the block.  For something like a sign, the floor co-ordinates also depend on which way the sign is facing.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,58 @@
+
+    /**
+     * Get the closest face of the block intersected by the entity's line of
+     * sight.  This may be null if the line of sight did not intersect this
```
Why would it ever be null? Wouldn't that preclude the concept of it intersecting?

desht replied:
getLineOfSight() returns a `List<BlockTargetResult>`: each member of that list will have a non-null getBlock() indicating that line of sight passed through that block, but only the last element could possibly have a non-null getTargetFace() and getTargetLocation() return (and it might not, if the line of sight doesn't intersect that block's hitbox either).

Wolvereness replied:
Still makes no sense, if the light of sight doesn't intersect the block, then the block wouldn't have been included in the first place.

Also does not explain what happens in a perfect 45-degree angle.

desht replied:
There are two separate considerations:
1. Did the line of sight pass through the block?
2. Did the line of sight actually intersect the block's contents (i.e. the volume enclosed by `block.getHitbox()`) ?

Consider the player targeting a stone wall, in front of which there is a carpet block (0.0625 blocks high).  If the player is looking at the wall behind the carpet, then the targeted block will be the stone block.  `getLineOfSight()` will return a list of BlockTargetResult objects, the last element of which is the stone block - that element will have non-null `getTargetFace()` and `getTargetLocation()` returns.  The preceding element will be the carpet block - its `getBlock()` method returns the carpet block, but the `getTargetFace()` and `getTargetLocation()` methods will return null, indicating that the player's line of sight did not in fact intersect the carpet's hit box - it passed over it.

This is a fairly significant feature - the old API willl simply consider the carpet block to be targeted, even when the player's line of sight is nowhere near the actual carpet.  (But the `useHitBoxes` parameter for the various targeting methods is provided in case plugins want this behaviour).

Regarding perfect 45-degree angles: nothing's perfect when it comes to floating-point representations.  The calculations will always end up on one side of the block boundary or other, and if the player's crosshairs are so finely positioned on a block boundary that a calculation becomes borderline, they won't be able to tell the difference either.  I'm not sure what's significant about 45 degrees anyway - either the player's line of sight intersects a plane, or it doesn't, regardless of the angle.

Update: having done more testing with my test plugin, having the crosshairs right on a boundary works fine.  The plugin always reports the same block as the client has highlighted with the targeting box.

desht replied:
As another note, the result from the (new) `getTargetBlock()` method will always be non-null; a BlockTargetResult object.  To see if the player actually has something targeted, check if either `getTargetedFace()` or `getTargetedLocation()` method of the BlockTargetResult return a non-null value.  I suppose a `boolean isBlockTargeted()` method could be added as a convenience/clarity method.

The `getBlock()` method returns the block being looked at; in the case of the player looking into thin air, for example, it will be the block that is `maxDistance` blocks along the line of sight (and in that case, both `getTargetedFace()` and `getTargetedLocation()` will return null).

Wolvereness replied:
> The preceding element will be the carpet block - its `getBlock()` method returns the carpet block, but the `getTargetFace()` and `getTargetLocation()` methods will return null, indicating that the player's line of sight did not in fact intersect the carpet's hit box - it passed over it.

That needs to be corrected. You are including the entire voxel (1 x 1 x 1) as calculated. You wouldn't give _null_ for an `AIR` block, would you? If you want to indicate the actual bounded area occupied by the block was not intersected, even if though voxel was, you need to define a flag.

Additionally, how would you know the carpet wasn't included because of its volume? Aren't you using a `Predicate<Block>`?

desht replied:
Maybe we're at cross-purposes here - the simple presence of the BlockTargetResult object in the List returned by getLineOfSight() indicates that the voxel (the 1x1x1 container) _was_ intersected by the entity's line of sight.

I'd know that the carpet wasn't included because:
1. It's not the last object in the list (so the entity's line of sight passed through the voxel, but not through the carpet itself)
2. The `getTargetedFace()` and `getTargetedLocation()` methods for the BlockTargetResult object return null.

There's a distinction to be drawn between blocks to be ignored because of transparency information (the Predicate) and blocks to be ignored because the line of sight doesn't intersect the block's hit box.

It's important to remember that unlike the old API, these methods are returning a targeting result record (or List thereof), not an actual Block.  The result record includes the block that was considered (`getBlock()`) and the results of considering that block (`getTargetedFace()` and `getTargetedLocation()`).  For `getLineOfSight()` and `getLastTwoTargetBlocks()`, the intermediate records along the line of sight will, by definition, be considered intermediate either due to the Predicate or because the block's hit box was not intersected; only the final record in the returned list _might_ have an intersection (and that depends on whether an intersection was found within the `maxDistance` parameter).

Regarding `AIR`: that's a rather special case, since AIR has no hit box at all.  However if you really want to target the nearest AIR block (and we're getting pretty contrived here), pass a `Predicate<Block>` which returns true for AIR, and a `useHitBoxes` parameter of false.

nightpool replied:
> You wouldn't give null for an AIR block, would you?
> How would you define the targeted face of an air block? Its a concept that minecraft doesn't use, and I'm guessing it doesn't have any code to back it up in the internals.

In fact, this is the first line of the lineOfSightIntersection implementation:

```
if (block.getType() == Material.AIR) {
    return new BlockTargetResult(block, null, null);
}
```

And you can get the behavior you want with regular blocks by passing useHitBoxes = false; it'll just give you the intersection point with the block's volume ("the entire voxel").

Wolvereness replied:
> It's important to remember that unlike the old API ...

Leave the old API out of this, unless we are specifically referring to a regression in features.

Giving _null_ for blockface as an indicator for if the voxel's occupied volume collided with the ray-trace is bad API-design, especially when there is no logical leap to defining the blockface from the voxel itself. Define me a situation where the blockface of intersection violates the blockface of the containing voxel - therein lies the only justification.

@Wolvereness (excuse my monologue response)

> Torches

Okay, so do we handle torch's BlockFace.EAST when the voxel's intersection was BlockFace.NORTH? The documentation sucks if that's the case.

So, is TargetFace defined by the voxel or the occupants?
- By voxel implies there should always exist a target-face, and overloading the meaning of _null_ is ridiculous
  - We have an opportunity to make API that doesn't suck
  - This is just another _gotcha_
- By occupants implies there should be both
  - Better documentation
  - Another target face for voxel-intersection

Oh, and if we're calculating occupant intersections, we need to make the predicate accept `BlockTargetResult` for indications, otherwise forcing the intersection math onto plugins.

nightpool replied:
Occupants. The target face for voxel-intersection can be found by passing
useHitBoxes = False

On Thu, May 29, 2014 at 8:37 AM, Wesley Wolfe notifications@github.comwrote:

> In src/main/java/org/bukkit/block/BlockTargetResult.java:
> 
> > -        this.face = face;
> > -        this.location = location;
> > -    }
> >   +
> > -    /**
> > -     \* Get the block which is intersected by the entity's line of sight.
> > -     *
> > -     \* @return the block
> > -     */
> > -    public Block getBlock() {
> > -        return block;
> > -    }
> >   +
> > -    /**
> > -     \* Get the closest face of the block intersected by the entity's line of
> > -     \* sight.  This may be null if the line of sight did not intersect this
> 
>  It's important to remember that unlike the old API ...
> 
> Leave the old API out of this, unless we are specifically referring to a
> regression in features.
> 
> Giving _null_ for blockface as an indicator for if the voxel's occupied
> volume collided with the ray-trace is bad API-design, especially when there
> is no logical leap to defining the blockface from the voxel itself. Define
> me a situation where the blockface of intersection violates the blockface
> of the containing voxel - therein lies the only justification.
> 
> @Wolvereness https://github.com/Wolvereness (excuse my monologue
> response)
> 
> Torches
> 
> Okay, so do we handle torch's BlockFace.EAST when the voxel's intersection
> was BlockFace.NORTH? The documentation sucks if that's the case.
> 
> So, is TargetFace defined by the voxel or the occupants?
> - By voxel implies there should always exist a target-face, and
>   overloading the meaning of _null_ is ridiculous
>   - We have an opportunity to make API that doesn't suck
>   - This is just another _gotcha_
>     - By occupants implies there should be both
>   - Better documentation
>   - Another target face for voxel-intersection
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Bukkit/Bukkit/pull/1059/files#r13180281
> .

Wolvereness replied:
@nightpool https://github.com/Bukkit/Bukkit/pull/1059/files#r13180369

> Occupants. The target face for voxel-intersection can be found by passing `useHitBoxes = False`

Sounds like the method itself is trying to overcome a shortcoming of the `Block` parameter in the `Predicate<Block>`. Why not change the method to take a `Predicate<BlockTargetResult>`, with said result able to indicate the type of intersection, and both `BlockFace` (_null_ only used for occupant's lack-of LOS intersection).

desht replied:
Yes, very definitely the occupants.  I'll accept the documentation could make it clearer, but I think this API actually makes sense: you ask what an entity is targeting, you get a record back.  The record includes the overall voxel, and (if there's an intersection) the face of the hitbox/voxel and the precise location of the intersection on the hitbox/voxel.  Hitbox/voxel being decided by the `useHitboxes` parameter.

If you look at a torch (so the torch's hit box is drawn on your client) and do a `player.getTargetBlock(null, 100, true)`, you'll get back a BlockTargetResult, where:
- `getBlock()` gives you the voxel for the torch;
- `getTargetFace()` gives you the face of the hit box that you are looking at;
- `getTargetLocation()` gives you the precise location where your line of sight intersects the face of the hit box (the face being what `getTargetFace()` returns).

If you pass `false` instead of `true` for `useHitboxes` then the above still applies, but for "hit box" read "voxel".

I don't get your comment about making the predicate accept a BlockTargetResult at all - what math would plugins need to do here?  All the math is done in the implementation - look at my CraftLivingEntity changes in the CB PR.

Wolvereness replied:
> look at my CraftLivingEntity changes in the CB PR

I'm evaluating the API - which (usually) needs to be designed independently of the implementation.

> If you pass `false` instead of `true` for `useHitboxes` then the above still applies, but for "hit box" read "voxel".

Why not just let the `Predicate<BlockTargetResult>` decide if they care between the hitbox and voxel?

desht replied:
That predicate is for deciding transparency; overloading it to be a) a transparency filter, b) a hitbox/voxel selector and c) a way of returning the results feels like trying to stuff a quart in a pint pot, so to speak.  I don't think reducing the number of method parameters aids usability here.

There's also the consideration that moving the `useHitBoxes` parameter into the predicate makes the method signature ambiguously similar to the old API: if you want to pass a null predicate (which IMHO should be a valid thing to do for the common case of "what am I looking at?"), it'll need an explicit cast to `(Predicate<BlockTargetResult>) null` to avoid confusion with the old `Set<Integer>` methods.  That's just irritating.

desht replied:
I should also mention that calculating by voxel is almost always the _wrong_ approach; it's aggravating as hell to be unable to target a wall behind a sign when your crosshairs are very clearly on the wall and not the sign.  The `useHitBoxes` parameter should generally be true; it's mainly there to provide backwards compatibility in case any plugin is depending upon the old functionality for whatever reason (the fact that it also fully disambiguates the method signatures is just a bonus).

desht committed:
Better Javadoc formatting for BlockTargetResult description

desht commented:
Argh, my git skills have failed.  I may need to close this PR and create a new one.

nightpool commented:
Just keep pushing commits. Travis or whoever merges this will squash before
committing.

On Wed, May 28, 2014 at 2:15 PM, desht notifications@github.com wrote:

> Argh, my git skills have failed. I may need to close this PR and create a
> new one.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/Bukkit/Bukkit/pull/1059#issuecomment-44451904
> .

desht commented:
@nightpool It's OK, I fixed it.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Represents the result of calling a method to determine blocks intersected
+ * by an entity's line of sight:
+ * <p><ul>
```
Much prefer the &lt;p&gt; to be its own line.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,57 @@
+ * <p><ul>
+ * <li>{@link LivingEntity#getLineOfSight(Predicate, int, boolean)}</li>
+ * <li>{@link LivingEntity#getTargetBlock(Predicate, int, boolean)}</li>
+ * <li>{@link LivingEntity#getLastTwoTargetBlocks(Predicate, int, boolean)}</li>
```
Too long.

Wolvereness commented:
If necessary, someone (myself?) can throw together a utility class that gives a `Predicate<Block>` from a set.

``` Java

public static final Function<Block, Material> TYPE_OF = new Function<Block, Material>() { 
    @Override
    public Material apply(Block block) {
        return block.getType();
    }
};

public static Predicate<Block> hasType(Collection<? extends Material> set) {
    return Predicates.compose(Predicates.in(set), TYPE_OF);
}
```

Yeah, there's not much to it...

desht committed:
Javadoc reformat; fit into 72-char width

desht committed:
Updated Block hitbox API following discussion with Wolvereness

Wolvereness commented:
Just noticed, `Predicate<Block>` should be `Predicate<? super Block>`.

This allows `Predicate<Object>` or `Predicate<Metadatable>` to function, which are logically perfectly valid predicates.

Wolvereness commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,64 @@
+ * <ul>
+ * <li>
+ *     {@link LivingEntity#getLineOfSight(Predicate, int, boolean)}
+ * </li>
```
Although we do need to start adding `</li>` tags to code, I think the best standard is as shown in the JDK 1.8 PR https://github.com/Bukkit/Bukkit-JavaDoc/pull/87/files#diff-9ab353926622cde8f0a2ee753b8138b1R223

desht committed:
Use Predicate<? super Block> for transparency parameter

desht committed:
Javadoc updates & clarifications

Wolvereness commented:
> I should also mention that calculating by voxel is almost always the _wrong_ 

I want the `Predicate` deciding if the raytrace should stop. All the parameters are meaningless if you allow the predicate to do just that.

Additionally, _null_ indicating specific behavior when it's supposed to be functional is a terrible idea.

desht commented:
Alright, focusing on just `getTargetBlock()` for now, you want a method like:

```
public BlockTargetResult getTargetBlock(Predicate<BlockTargetResult> predicate);
```

Yes?

So the plugin is responsible for providing a predicate which returns false when the raytrace should stop, and the method returns the targeting result at that block.  So the predicate has to decide if a) the block should be ignored completely (transparent), b) whether to stop due to the distance of the block from the entity, and c) whether to use the block's hit box or its full voxel volume to determine intersections.

I could be missing something here, but that's moving a lot of work to the plugin which IMHO should be done within the method.  I can see adding the distance check to the predicate (although it makes the whole thing more verbose - forcing the plugin to make an explicit distance comparison in the predicate instead of passing an int parameter), but the hitbox/voxel decision is orthogonal to telling the raytrace whether or not to continue; it doesn't belong in a "should I stop here?" predicate at all.

If you could provide an example of how that would be used to translate, say:

``` java
Block b = player.getTargetBlock(new Predicate<Block>() {
  @Override
  public boolean apply(Block b) {
    return b.getType() == Material.AIR || b.getType() == Material.GLASS;
  }
}, 120, true);
```

then maybe I'll understand better.

---

I've also been thinking about this separately, and it occurred to me that I can dispense with the BlockTargetResult class completely, simplifying the method to:

``` java
public Block getTargetBlock(Predicate<Block> transparent, int maxDistance, boolean useHitBoxes);
```

This method just returns the block that the player is targeting, if any, which is the most common case.  In the rarer cases where plugins need to know the targeted block face and/or intersection point, two extra methods will be provided:

``` java
public BlockFace getTargetBlockFace(Block block, boolean useHitBox);
public Location getTargetLocation(Block block, boolean useHitBox);
```

where `block` will usually be the result of a previous call to `getTargetBlock()`.

---

Finally, if we want to avoid using a null predicate to mean "treat AIR as transparent", we'll need a static utility method to provide a suitable predicate (`isAir()` ?), which can go in a class with your suggested `hasType()` method.  org.bukkit.util.BlockPredicates maybe?  (or should it go in org.bukkit.block?)

Wolvereness commented:
All of this actually makes me think just returning an `Iterable<BlockTargetResult>` would be best, with suitable wrappers (as utility) to stop at the last block, or stop at certain transparencies, etc.

Although, we already have generic ray-tracing, don't we?

desht commented:
Yes, we have BlockIterator already (and my implementation still makes use of that - yeah, I know, don't talk about implementation...).

I'm becoming more & more convinced that `BlockTargetResult` is an unnecessary addition (and I'm pretty sure `getLastTwoTargetBlocks()` isn't very useful either).

Not sure how a `Iterable<BlockTargetResult` would work.  Perhaps for `getLineOfSight()` (although I'm still trying to figure out how it would fit) but definitely not for `getTargetBlock()`.

turt2live commented on src/main/java/org/bukkit/block/Block.java:
```diff
@@ -381,4 +382,42 @@
+     * @param floor Location in which to store the hit box's floor location
+     * @param size Vector in which to store the hit box's size
+     */
+    void getHitBox(Location floor, Vector size);
```
~~Any particular reason this doesn't return a `Vector` ?~~ I know there was mention of a `Tuple` however that conversation seems to have stopped at "not in Bukkit". 

Edited: Just looked at implementation and misread the docs.

desht replied:
Yep, this was suggested by @Wolvereness  as a way to get both the floor and size of the block's hitbox in a single call.

libraryaddict replied:
Couldn't you return a Entry? Such as new HashMap.SimpleEntry(size, hitbox);

desht replied:
@libraryaddict no, because this isn't a mapping; it's just a pair of values.

turt2live commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Represents the result of calling the following methods to determine blocks
+ * or block hit boxes intersected by an entity's line of sight:
+ * <ul>
```
~~These JavaDocs are slightly off for formatting.~~

I've been made aware that this is actually correct as is.

turt2live commented on src/main/java/org/bukkit/block/BlockTargetResult.java:
```diff
@@ -0,0 +1,71 @@
+    private final BlockFace face;
+    private final Location location;
+
+    public BlockTargetResult(Block block, BlockFace face, Location location) {
```
No validation?

desht replied:
No, but I want to get rid of this class anyway.

turt2live commented:
@desht I can see the `getLineOfSight()` method returning an iterable for something like this:

``` java
Iterable<BlockTargetResult> los = player.getLineOfSight(....);
for(BlockTargetResult result : los){
    // Magic
}
```

As for the usefulness of `BlockTargetResult` - I would agree with it becoming more and more useless as this PR has moved along. I'm not sure what options you have considered for replacements, so sharing those would be appreciated ;) 

As it currently stands, I'm impressed by the progress thus far and look forward to seeing this in the API ;)

desht commented:
@turt2live here are my thoughts at this point:
- I think the BlockTargetResult class was a mistake and I want to get rid of it.
- I also think the `getLastTwoTargetBlocks()` method is probably unnecessary; the only use I can think is to get the targeted block face (via `block.getRelative(Block, Block)`) but 1) that's unreliable if the player is really close to the targeted block and 2) the new API can handle that more cleanly with an explicit `BlockFace getTargetFace(Block)` call.

So, I propose to change the PR to add `Block getTargetBlock(Predicate<Block>, int, boolean)` and `List<Block> getLineOfSight(Predicate<Block>, int, boolean)` methods which simply return a block or blocks, not a BlockTargetResult.  This is by far the most common case.  In the less common case where a plugin needs to know either the targeted block face or the precise targeted Location, separate `BlockFace getTargetFace(Block)` and `Location getTargetLocation(Block)` methods can be used, where the passed parameter would typically be the return value of a previous `getTargetBlock()` call.

There's also the issue of adding some utility predicates as proposed by @Wolvereness at a couple of places in this thread, e.g. https://github.com/Bukkit/Bukkit/pull/1059#issuecomment-44473217 .  I'd like some input on what should be in this class, and where it should go (I suggest a static `BlockPredicates` class, either in `org.bukkit.block` or `org.bukkit.util`).

turt2live commented:
Getting rid of that class is a good first step. Implementing some kind of Predicate class is likely required and I'll wait to see what comes of that step :)

desht committed:
removed BlockTargetResult, updated LivingEntity interface methods

desht commented:
OK, changes made:
- BlockTargetResult is gone
- New `getTargetBlock()` and `getLineOfSight()` methods now return a `Block` and `List<Block>` respectively.
- New `getLastTwoTargetBlocks()` method is gone.  It simply doesn't seem useful.
- Added further two new methods to LivingEntity: `BlockFace getTargetBlockFace(Block block, boolean useHitbox)` and `Location getTargetLocation(Block block, boolean useHitbox)`
  Implementation also updated to match.  Tested everything compiled, updated my test plugin (see link in original description), and verified results are still as expected; all appears well.

me4502 commented:
Out of curiosity, how accurate is this to the client?

desht commented:
@me4502 extremely accurate, as far as I can tell.  Hitboxes as displayed on the client match up with the hitboxes calculated by the server very well indeed.

The one exception is stairs, where the hitbox returned by the server is just wrong (but it's pretty broken in vanilla too - check Mojang's JIRA for bug reports related to stairs and hitboxes).  But stairs are a weird case anyway - a properly accurate hitbox wouldn't be possible with a single AABB.
